hail/src/main/scala/is/hail/expr/types/physical/PContainer.scala

```diff
@@ -254,36 +260,51 @@ abstract class PContainer extends PIterable {
     }
   }
 
-  def checkedConvertFrom(mb: EmitMethodBuilder, r: Code[Region], value: Code[Long], otherPT: PType, msg: String): Code[Long] = {
-    val otherPTA = otherPT.asInstanceOf[PArray]
-    assert(otherPTA.elementType.isPrimitive)
-    val oldOffset = value
-    val len = otherPTA.loadLength(oldOffset)
-    if (otherPTA.elementType.required == elementType.required) {
-      value
-    } else {
-      val newOffset = mb.newField[Long]
-      Code(
-        newOffset := allocate(r, len),
-        stagedInitialize(newOffset, len),
-        if (otherPTA.elementType.required) {
-          // convert from required to non-required
-          Code._empty
-        } else {
-          //  convert from non-required to required
-          val i = mb.newField[Int]
+  def ensureNoMissingValues(mb: EmitMethodBuilder, sourceOffset: Code[Long], sourceType: PContainer, onFail: Code[_]): Code[Unit] = {
+    if(sourceType.elementType.required) {
+      return Code._empty
+    }
+
+    val i = mb.newLocal[Long]
+    Code(
+      i := PContainer.nMissingBytes(sourceType.loadLength(sourceOffset)),
+      Code.whileLoop(i > 0L,
+        (i >= 8L).mux(
           Code(
-            i := 0,
-            Code.whileLoop(i < len,
-              otherPTA.isElementMissing(oldOffset, i).orEmpty(Code._fatal(s"${msg}: convertFrom $otherPT failed: element missing.")),
-              i := i + 1
-            )
+            i := i - 8L,
+            Region
+              .loadLong(sourceOffset + sourceType.lengthHeaderBytes + i)
+              .cne(const(0.toByte))
+              .orEmpty(onFail)
+          ),
+          Code(
+            i := i - 1L,
+            Region
+              .loadByte(sourceOffset + sourceType.lengthHeaderBytes + i)
+              .cne(const(0.toByte))
```

Comment:
@akotlar @johnc1231, am I reading this correctly that in a datatype that has one missing bit, say `struct<x: int32>`, this will compare the entire byte allocated to missingness to the 0 byte? I'm not sure we guarantee that the unused excess bits in the last missingness byte are zero'd.

Reply:
When John stagedInitializes, the entire missing bit array is 0’d